### PR TITLE
Introduces better VaRA documentation

### PR DESCRIPTION
**Pull request by vulder**

### CODE REVIEW

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
+Feature Performance Analysis
+============================
+
+VaRA's feature performance analysis is build to make the impact of software features measurable.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
VaRA's feature performance analysis is built to make the impact of software features measurable.
```

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+By only specifying where control variables are, our analysis pipeline detects feature-dependent code parts and instruments them for further run-time analysis.
+
+
+How to setup and use the feature performance analysis
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
How to set up and use the feature performance analysis
```

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+-----------------------------------------------------
+
+Setting up our analysis is simple and can be done in a few steps.
+
+1) Setup VaRA by following our :ref:`setup guide<Build VaRA with vara-buildsetup>`.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
1) Set up VaRA by following our :ref:`setup guide<Build VaRA with vara-buildsetup>`.
```

File: docs/source/research_tool_docs/vara/using.rst [hunk 4]
```diff
@@ -0,0 +1,30 @@
+**************************
+How to use VaRA (Overview)
+**************************
+
+Overall, VaRA is build for two different usage scenarios:
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/using.rst]:
```suggestion
Overall, VaRA is built for two different usage scenarios:
```

File: docs/source/research_tool_docs/vara/vara.rst [hunk 5]
```diff
@@ -1,4 +1,14 @@
-.. figure:: VaRA_Logo_31.png
+****
+VaRA
+****
+
+VaRA is an analysis framework that enables users to build static and dynamic analyses for analyzing high-level concepts, using advanced compiler and analysis technology in the background.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/vara.rst]:
```suggestion
VaRA is an analysis framework that enables users to build static and dynamic analyses for analyzing high-level concepts using advanced compiler and analysis technology in the background.
```

File: docs/source/research_tool_docs/vara/vara.rst [hunk 6]
```diff
@@ -1,4 +1,14 @@
...
+Our goal is to enable the user to build these analyses by only focusing on the highlevel conceptual information that should be analyzed, without worrying about low-level details, such as, building complicated compiler modifications or configuring precise but difficult to use static analyses.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/vara.rst]:
```suggestion
Our goal is to enable the user to build these analyses by only focusing on the high-level conceptual information that should be analyzed without worrying about low-level details, such as building complicated compiler modifications or configuring precise but difficult to use static analyses.
```

File: docs/source/research_tool_docs/vara/vara.rst [hunk 7]
```diff
@@ -1,4 +1,14 @@
...
+
+The figure below, gives a rough overview of VaRA and the analysis process.
+In general, VaRA and our modified clang compiler take in source code together with highlevel conceptual information and either directly analyze it, using various static analyses, or generate an instrumened binary, which can run different dynamic analyses.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/vara.rst]:
```suggestion
In general, VaRA and our modified clang compiler take as input source code together with high-level conceptual information and either directly analyze it, using various static analyses, or generate an instrumented binary, which can run different dynamic analyses.
```

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    CC=$PATH_TO_VARA_ROOT/tools/VaRA/bin/clang
+    CXX=$PATH_TO_VARA_ROOT/tools/VaRA/bin/clang++
+
+
+3) Set the feature specific compile flags as (`C_FLAGS` or `CXX_FLAGS`).
+************************************************************************
+
+Enable the automatic feature detection (`-fvara-feature`) and select the desired instrumentation code (`-fvara-instr=`).
+The different instrumentation options can be found here :ref:`Instrumentations`.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
The different instrumentation options can be found :ref:`here<Instrumentations>`.
```

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+.. code-block:: console
+
+    CXX_FLAGS="$CXX_FLAGS -fvara-feature -fsanitize=vara -fvara-instr=trace_event"
+
+
+(Recommended) Configure your project to use link-time optimization (LTO) for more precise analysis results.
+Without LTO, only within translation unit analysis is possible, hence, cross translation unit feature usage is not possible.
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
Without LTO, the analysis can only run within a translation unit, hence, features that are used across different translation units can not be correctly analyzed.
```

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+.. code-block:: console
+
+    CXX_FLAGS="$CXX_FLAGS -fvara-fm-path=/path/to/the/feature_model.xml"
+
+
+Or, specify feature location within your code direclty.
+
+.. code-block:: cpp
+
+    bool MyFeatureVariable __attribute__((feature_variable("FEATURE_NAME"))) = false /*default*/;
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]: A bit more explanation here would be good:
- what does the feature model do?
- Where does the feature model come from and how do I get one for my project?
- what happens if I mark a variable this way?

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+.. code-block:: console
+
+    export VARA_TRACE_FILE=my_little_tracefile.json
+
+
+Instrumentations
+----------------
+
+* `print`: Print entry/exit messages when entring a feature specific code
+* `clock`: Add hw-clock based measurements that determine the time spend in a feature
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
* `clock`: Add hw-clock based measurements that determine the time spent in a feature
```

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+* `trace_event`: Add trace event markers that generate catapult (trace event format) files
+* `instr_verify`: Add verifier instrumentation that check if regions are correctly opened/closed
```
Comment (boehmseb) [inline docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]:
```suggestion
* `instr_verify`: Add verifier instrumentation that checks if feature regions are correctly opened/closed
```

**Review by boehmseb — CHANGES_REQUESTED** [review]
Overall a good addition to improve the VaRA documentation. I found some spots where the formulation could be a bit clearer or where some additional information would be nice. And of course some typos ;)

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst (hunk 10 above)
Comment (vulder) [reply docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]: Ok now?

**Review by boehmseb — APPROVED** [review]

File: docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst (hunk 10 above)
Comment (boehmseb) [reply docs/source/research_tool_docs/vara/usages/feature_performance_analysis.rst]: Yes

**Pull request merged**